fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/bitmap/ComposedBitmapIndexAggregate.java

```diff
@@ -319,9 +321,26 @@ private static boolean separateGroupFilters(@Nonnull QueryComponent filter,
                 return Optional.empty();
             }
             // Splice the index's key between the common grouping key and the position.
-            GroupingKeyExpression groupKeyExpression = (GroupingKeyExpression)indexAggregateFunction.getOperand();
-            GroupingKeyExpression fullKey = Key.Expressions.concat(groupKeyExpression.getGroupingSubKey(), indexKey, groupKeyExpression.getGroupedSubKey())
-                    .group(groupKeyExpression.getGroupedCount());
+            // The simplest place is directly before the position.
+            // But if part of the group is a nested concat, breaking that up would need support in QueryToKeyMatcher.
+            // Moreover, the caller needs to have arranged for a compatible index to exist, which requires new support.
+            // (https://github.com/FoundationDB/fdb-record-layer/issues/1056)
+            final GroupingKeyExpression groupKey = (GroupingKeyExpression)indexAggregateFunction.getOperand();
+            int lastKeySize = 1;
+            if (groupKey.getWholeKey() instanceof ThenKeyExpression) {
+                ThenKeyExpression thenKey = (ThenKeyExpression)groupKey.getWholeKey();
+                lastKeySize = thenKey.getChildren().get(thenKey.getChildren().size() - 1).getColumnSize();
+            }
+            final ThenKeyExpression splicedKey;
+            if (lastKeySize > groupKey.getGroupedCount()) {
```

Comment:
Can this still end up splicing up a nested concat? I'm thinking of something like if `groupKey.getGroupedCount() == 2` and `groupKey` is something like `concat(nest(a, concat(b, c)), d)`?

Reply:
The new version will always take whole children of a `Then`.